File: ros2-for-beginners/ros2_ws/src/my_py_pkg/my_py_pkg/add_two_ints_client_no_oop.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


def main(args=None):
    rclpy.init(args=args)
    node = Node("add_two_ints_no_oop")

    """
    create a client
        takes in 2 parameters 
            1. service type 
            2. service name
    """
    client = node.create_client(AddTwoInts, "add_two_ints")

    # wait for server
    while not client.wait_for_service(1.0):
        node.get_logger().warn("waiting for service add two ints...")

    # exit while
    node.get_logger().info("\n<-- server connected successfully -->")

    # create a request object using AddTwoInts
    request = AddTwoInts.Request()
    request.a = int(input('enter number 1 :'))
    request.b = int(input('enter number 2 :'))

    """
    call function types 
        1. call
        2. call_async

        wait for the call to succeed and the future has a value from the response 
        - spin the node until the future gets a value
            spin_until_future_complete 
                takes 2 arguements 
                    1. node -> for referring to the currently created node 
                    2. future -> for waiting until it has a value

            in the future the result object contains the actual response from the server
    """
    future = client.call_async(request)
    rclpy.spin_until_future_complete(node, future)

    # take out result from the future using try except
    try:
        response = future.result()
        node.get_logger().info(f"{request.a} + {request.b} = {response.sum}")
    except Exception as e:
        node.get_logger().error(f"service call failed : {e}")

    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2-for-beginners/ros2_ws/src/my_cpp_pkg/src/add_two_ints_client_no_oop.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp" 

#include <cstdlib>

int main(int argc, char **argv){
    rclcpp::init(argc,argv);


    if(argc !=3) {
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"),"usage : add_two_ints_client_no_oop X Y");
        return 1;
    }
    
    auto node = std::make_shared<rclcpp::Node>("add_two_ints_client_no_opp");
    /*
        create a client 
        send request 
        wait for the response 
    */
    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
    while(!client->wait_for_service(std::chrono::seconds(1))){
        RCLCPP_WARN(node->get_logger(),"cpp client waiting for add to int server");
    }

    RCLCPP_INFO(node->get_logger(),"<-- service available -->");

    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    request->a = atoll(argv[1]);
    request->b = atoll(argv[2]);

    auto future = client->async_send_request(request);
    if(rclcpp::spin_until_future_complete(node,future) == rclcpp::FutureReturnCode::SUCCESS){
        RCLCPP_INFO(node->get_logger(),"%ld + %ld = %ld",request->a,request->b,future.get()->sum);
    }
    else{
        RCLCPP_ERROR(node->get_logger(),"Failed to call service add_two_ints");
    }

    
    rclcpp::shutdown();
    return 0;
}